### PR TITLE
Implement polygon modes in geometry shader

### PR DESCRIPTION
**Pull request by JayFoxRox**
Replaces https://github.com/espes/xqemu/pull/58

This prepares us for doing most oldschool rendering specific tasks in shaders:
First I added QUAD_STRIP and POLYGON modes which should work in all polygon modes [POINT, LINE, FILL].
I then refactored and extend it to keep all polygon mode and primitive mode related stuff in a single function.
It should prepare us for two-sided-polygon modes and it will remove any ambiguity with primitive modes and geometry shaders.
Furthermore I decided to stop inlining the common GLSL code in the C portion of the code. GLSL has a good rules about constants and the GPU driver is free to inline the functions itself.

I unit-tested every TRIANGLE\* and QUAD\* primitive type in LINE and FILL mode and ran some games. I haven't come accross any problems yet.
Polygon winding might be wrong so I'll check that again.

(I plan to write a blog post about my experience with recreating those old GL features in the GS to make things a little clearer)

tl;dr: All draw-calls will go through the same shader generation function now. We should have support for all primitive types [except for patches!] in every supported polygon mode.

### CODE REVIEW

File: hw/xbox/nv2a.c [hunk 1]
```diff
@@ -1636,7 +1615,8 @@ typedef struct PGRAPHState {
     hwaddr dma_vertex_a, dma_vertex_b;
 
     unsigned int primitive_mode;
-    GLenum gl_primitive_mode;
+    unsigned int polygon_front_mode;
+    unsigned int polygon_back_mode;
```
Comment (espes) [inline hw/xbox/nv2a.c]: are these unused?

File: hw/xbox/nv2a_shaders.c [hunk 2]
```diff
@@ -23,50 +23,166 @@
...
-        "noperspective in VertexData v_vtx[];\n");
-    qstring_append(s,
-        "noperspective out VertexData g_vtx;\n");
-    qstring_append(s, "\n");
 
-    qstring_append(s, "void main() {\n");
+    /* Handle LINE and FILL mode */
+    const char *layout_in;
+    const char *layout_out;
+    const char *body;
```
Comment (espes) [inline hw/xbox/nv2a_shaders.c]: nit: init these to NULL

**Comment by espes** [comment]
Looks good! Feel free to merge this addressing those nits. Thanks!

File: hw/xbox/nv2a.c [hunk 3]
```diff
@@ -5165,7 +5149,7 @@ static void pgraph_method(NV2AState *d,
 
                 pgraph_bind_vertex_attributes(d, pg->draw_arrays_max_count,
                                               false, 0);
-                glMultiDrawArrays(pg->gl_primitive_mode,
+                glMultiDrawArrays(pg->shader_binding->gl_primitive_mode,
                                   pg->gl_draw_arrays_start,
```
Comment (espes) [inline hw/xbox/nv2a.c:5153]: Put in an assert(pg->shader_binding) above.

File: hw/xbox/nv2a_shaders.c (hunk 2 above)
Comment (JayFoxRox) [inline hw/xbox/nv2a_shaders.c]: I thought about it but I thought I'd rather get a warning if they are set and not used
or
warning if they are not set but used
[which the compiler should do for the current code]

Whenever `layout_in`, `layout_out` or `body` are used the code should reach the geometry shader generator. If they are not set you should `return NULL` before.

However, I'll probably init to NULL to make sure that the assertions before the shader gen. will work [which effectivly turns compiler warnings into assert-cases]

File: hw/xbox/nv2a.c (hunk 3 above)
Comment (JayFoxRox) [inline hw/xbox/nv2a.c:5153]: Okay.

File: hw/xbox/nv2a.c (hunk 1 above)
Comment (JayFoxRox) [inline hw/xbox/nv2a.c]: Yes, moved those to ShaderState and forgot about this probably.
